### PR TITLE
[Woo POS][non-simple products] Add tests for Variations data source and remove virtual filtering

### LINKED ISSUE #13189: [Woo POS][non-simple products] Add tests for Variations data source and remove virtual filtering
The objective of this task is to enhance test coverage by adding unit tests for the `WooPosVariationsDataSource` class. Additionally, the filtering logic for variations will be moved into this class. As part of this task, virtual variations will also be included, aligning with the ongoing discussion.

### PR DESCRIPTION
**Pull request by AnirudhBhat**
Closes: #13189 

### Description

This PR:
1. Adds unit tests for `WooVariationsDataSource`
2. Moves filtering logic to `WooVariationsDataSource`
3. Remove virtual check in the filtering logic as we are leaning towards including virtual products and variations as well.

#### In Variations screen
1. Navigate to POS (more menu -> POS)
2. Click on any variable product
3. Ensure the filter is applied properly (downloadable and variations with no price should not be shown)
4. Ensure no items are removed after they are displayed

#### In Products screen
1. Navigate to POS (more menu -> POS)
2. Ensure the filter is applied properly (virtual products must be shown)

### Testing information

Same as "Steps to reproduce"

### Images/gif

https://github.com/user-attachments/assets/350cdfd4-6119-4692-9000-ce1827e03cc6

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/di/VariationsCacheModule.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import dagger.Module
+import dagger.Provides
+import dagger.hilt.InstallIn
+import dagger.hilt.components.SingletonComponent
+import javax.inject.Singleton
+
+@Module
+@InstallIn(SingletonComponent::class)
+object VariationsCacheModule {
+    private const val VARIATION_CACHE_MAX_SIZE = 50
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/di/VariationsCacheModule.kt]: np: if you have `*Variations*LRUCache`, a specific LRUCache for variations, then I probably its size should be inside, not by someone from outside

And the same for generic types. If it's specific for variations, then why it's generically typed?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/di/VariationsCacheModule.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import com.woocommerce.android.ui.woopos.home.items.variations.VariationsLRUCache
+import dagger.Module
+import dagger.Provides
+import dagger.hilt.InstallIn
+import dagger.hilt.components.SingletonComponent
+import javax.inject.Singleton
+
+@Module
+@InstallIn(SingletonComponent::class)
+object VariationsCacheModule {
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/di/VariationsCacheModule.kt]: The naming doesn't follow the naming convention, and we pollute the naming space of the main app with that

The same for `DataStoreModule`

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/variations/WooPosVariationsDataSourceTest.kt [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+
+        // WHEN
+        val result = sut.loadMore(productId)
+
+        // THEN
+        // Validate loadMore failure
+        assertThat(result.isFailure).isTrue()
+        assertThat(result.exceptionOrNull()).isEqualTo(exception)
+
+        // Validate cache remains unchanged
```
Comment (kidinov) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/variations/WooPosVariationsDataSourceTest.kt]: Are those GPT-generated comments? I'd propose removing them - just make the tests longer and bring a potential mismatch between the code and the comments. And Imo don't bring any value

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsDataSource.kt [hunk 4]
```diff
@@ -88,3 +87,7 @@ sealed class FetchResult<out T> {
     data class Cached<out T>(val data: T) : FetchResult<T>()
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsDataSource.kt]: Is there any need to have it generically typed? It looks like that it's always `List<ProductVariation>`

**Review by kidinov — APPROVED** [review]
Looks and tests good. I left a few comments, please take a look

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsDataSource.kt (hunk 4 above)
Comment (AnirudhBhat) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsDataSource.kt]: Not really required. We would if we decide to abstract it and start using it for more product types.

https://github.com/woocommerce/woocommerce-android/pull/13190/commits/7edd39f7e8abc43a1ce772b24c4e2415c57048be

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/variations/WooPosVariationsDataSourceTest.kt (hunk 3 above)
Comment (AnirudhBhat) [reply WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/variations/WooPosVariationsDataSourceTest.kt]: [Remove unnecessary comments](https://github.com/woocommerce/woocommerce-android/pull/13190/commits/39f15f4f75bdc3fdb173cd5e6211d9093cef7615)

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/di/VariationsCacheModule.kt (hunk 2 above)
Comment (AnirudhBhat) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/di/VariationsCacheModule.kt]: Removed the file and renamed `DataStoreModule`: https://github.com/woocommerce/woocommerce-android/pull/13190/commits/e25c71528d729bb7c0b327026a297ca1b9215a31  https://github.com/woocommerce/woocommerce-android/pull/13190/commits/4cfd0ddfd47f72b5990ff602b2700a78b2f8face

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/di/VariationsCacheModule.kt (hunk 1 above)
Comment (AnirudhBhat) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/di/VariationsCacheModule.kt]: Done here: https://github.com/woocommerce/woocommerce-android/pull/13190/commits/afcbd23a5acb9079e5e0fb9cd20f7b7c404aba21

**Pull request merged**